traffic_ops/traffic_ops_golang/ats/headerrewrite.go

```diff
@@ -0,0 +1,201 @@
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	cdnName, userErr, sysErr, errCode := getCDNNameFromNameOrID(inf.Tx.Tx, inf.Params["cdn-name-or-id"])
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	text, err := headerComment(inf.Tx.Tx, "CDN "+cdnName)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting hdr_rw_mid_xml-id.config text: "+err.Error()))
+		return
+	}
+
+	ds, err := getDeliveryService(inf.Tx, inf.Params["xml-id"])
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting hdr_rw_mid_xml-id.config text: "+err.Error()))
+		return
+	}
+
+	maxOriginConnections := *ds.MaxOriginConnections
+
+	dsType, err := deliveryservice.GetDeliveryServiceType(*ds.ID, inf.Tx.Tx)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting hdr_rw_mid_xml-id.config text: "+err.Error()))
+		return
+	}
+	usesMids := dsType.UsesMidCache()
+
+	// write a header rewrite rule if maxOriginConnections > 0 and the ds DOES use mids
+	if maxOriginConnections > 0 && usesMids {
+		dsOnlineMidCount, err := getOnlineDSMidCount(inf.Tx, *ds.ID)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting ds server count: "+err.Error()))
+			return
+		}
+		maxOriginConnectionsPerMid := int(math.Round(float64(maxOriginConnections) / float64(dsOnlineMidCount)))
+		if ds.MidHeaderRewrite == nil {
+			text += "cond %{REMAP_PSEUDO_HOOK}\nset-config proxy.config.http.origin_max_connections " + strconv.Itoa(maxOriginConnectionsPerMid) + " [L]"
+		} else {
+			text += "cond %{REMAP_PSEUDO_HOOK}\nset-config proxy.config.http.origin_max_connections " + strconv.Itoa(maxOriginConnectionsPerMid) + "\n"
+
+		}
+	}
+
+	// write the contents of ds.MidHeaderRewrite to hdr_rw_mid_xml-id.config replacing any instances of __RETURN__ (surrounded by spaces or not) with \n
+	if ds.MidHeaderRewrite != nil {
+		var re = regexp.MustCompile(`\s*__RETURN__\s*`)
+		text += re.ReplaceAllString(*ds.MidHeaderRewrite, "\n")
+	}
+
+	text += "\n"
+
+	w.Header().Set("Content-Type", "text/plain")
+	w.Write([]byte(text))
+}
+
+func getDeliveryService(tx *sqlx.Tx, xmlId string) (tc.DeliveryServiceNullable, error) {
+	qry := `SELECT id, cdn_id, max_origin_connections, edge_header_rewrite, mid_header_rewrite FROM deliveryservice WHERE xml_id = $1`
+	ds := tc.DeliveryServiceNullable{}
+	if err := tx.QueryRow(qry, xmlId).Scan(&ds.ID, &ds.CDNID, &ds.MaxOriginConnections, &ds.EdgeHeaderRewrite, &ds.MidHeaderRewrite); err != nil {
+		return tc.DeliveryServiceNullable{}, err
+	}
+	return ds, nil
+}
+// get the count of "online" edges assigned to a delivery service
+func getOnlineDSEdgeCount(tx *sqlx.Tx, dsID int) (int, error) {
```

Comment:
Nitpick: GoDoc should be `// getOnlineDSEdgeCount gets the count of online or reported edges assigned to a delivery service.`